### PR TITLE
fixed group detection bug for LDAP prune

### PR DESCRIPTION
**Pull request by stevekuznetsov**
fixes https://bugzilla.redhat.com/show_bug.cgi?id=1291607

@deads2k PTAL

### CODE REVIEW

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go [hunk 1]
```diff
@@ -26,6 +26,10 @@ func (l *GroupBasedDetector) Exists(ldapGroupUID string) (bool, error) {
 		return false, err
 	}
 
+	if group == nil {
+		return false, nil
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: @liggitt amazingly it is possible that when doing a specific search for a specific DN in LDAP we can get nothing back and *not* `NoSuchObject`

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go [hunk 2]
```diff
@@ -76,13 +80,13 @@ type CompoundDetector struct {
 }
 
 func (l *CompoundDetector) Exists(ldapGrouUID string) (bool, error) {
-	conclusion := false
+	conclusion := true
```
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: No.  If you don't have any locators, the result should be `false`, not `true`.

File: test/extended/ldap_groups.sh [hunk 3]
```diff
@@ -229,5 +229,14 @@ for (( i=0; i<${#schema[@]}; i++ )); do
 	openshift ex prune-groups --sync-config=sync-config.yaml --confirm
 	compare_and_cleanup valid_all_ldap_sync_prune.txt
 
+	# special-case test for ad-extended
```
Comment (deads2k) [inline test/extended/ldap_groups.sh]: move this outside of the loop.

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: Add a test that makes sure that no locators results in a `false` to make sure this never slips.

File: test/extended/ldap_groups.sh (hunk 3 above)
Comment (stevekuznetsov) [inline test/extended/ldap_groups.sh]: Doing that requires a lot of set-up that is done already in the loop... why can't we special-case it in the loop?

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go (hunk 1 above)
Comment (liggitt) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: what does the LDAP query look like for the DN lookup?

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go (hunk 2 above)
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: I don't think we should allow a `CompoundDetector` that has no constituent detectors. What would that be useful for?

**Comment by stevekuznetsov** [comment]
@deads2k comments addressed
as we are not doing cleanup and instead stomping on other temp files on startup I realized there is no extra work to be done.

[test][extended:ldap_groups]

**Comment by deads2k** [comment]
lgtm pending tests.

@stevekuznetsov update bugzilla with a reference to this pull.

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: > what does the LDAP query look like for the DN lookup?

@stevekuznetsov answer this to make sure that it looks sane before merge.

**Comment by stevekuznetsov** [comment]
@deads2k Bugzilla has the update already

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go (hunk 1 above)
Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: We haven't set the scope to base object only, that makes sense. We should be, however. I'll look into it.

Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: I think there may be larger problems. In either case, `QueryForUniqueEntry`, when it retrieves nothing, it should throw an `ErrorEntryNotFound`. Which the detector should recognize as meaning the thing it was looking for doesn't exist.

Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: Actually, never mind all of that. I can't read. The search request is:
```
&{cn=group1,ou=groups,ou=adextended,dc=example,dc=com 0 0 0 0 false (objectClass=*) [cn dn] []}
```
which translates to
```
&SearchRequest{
    BaseDN:       "cn=group1,ou=groups,ou=adextended,dc=example,dc=com",
    Scope:        ScopeBaseObject,
    DerefAliases: NeverDerefAliases,
    SizeLimit:    None,
    TimeLimit:    None,
    TypesOnly:    false,
    Filter:       "(objectClass=*)",
    Attributes:   ["cn", "dn"],
    Controls:     none,
}
```

I think everything is working as expected now.

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go [hunk 4]
```diff
@@ -76,13 +80,17 @@ type CompoundDetector struct {
...
+	}
+
+	conclusion := true
 	for _, locator := range l.locators {
 		opinion, err := locator.Exists(ldapGrouUID)
 		if err != nil {
 			return false, err
 		}
-		conclusion = conclusion || opinion
+		conclusion = conclusion && opinion
```
Comment (liggitt) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: can't we return false early if one returns false?

**Comment by stevekuznetsov** [comment]
@deads2k all green

File: pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go (hunk 4 above)
Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: > can't we return false early if one returns false?

Doc on the type guarantees an error in one errors.  I'm ok with forcing all successful inquiries to be sure about the decision

Comment (stevekuznetsov) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: If I recall correctly @deads2k and I decided we want all of the locators to run to ensure that no errors are encountered.

Comment (deads2k) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: > If I recall correctly @deads2k and I decided we want all of the locators to run to ensure that no errors are encountered.

You wanted, I had no strong opinion.

Comment (liggitt) [inline pkg/cmd/experimental/syncgroups/groupdetector/groupdetector.go]: ok

**Comment by deads2k** [comment]
[merge]

**Pull request merged**